### ISSUE TITLE
Balanced Pool

### ISSUE THREAD

**Issue by ronag** [mcollina/undici#310]
One problem with the current pool implementation is that if a lot of requests gets dispatched at the same time, it assumes that all pipelines will have the same throughput, however that is not necessarily true and cause queue imbalance.

When all clients are busy, queue requests and list to an event e.g. `'free'` which indicates that a client is no longer busy.

**Comment by mcollina** [comment]
Yes! One of the key strategy is to keep connections "wide and busy" by putting a lot of throughput through them. Balancing this something pseudo-optimal that does not increase too much the waiting time requires a lot more maths that I like to do (I left academia for a reason after all).

**Comment by delvedor** [comment]
A good solution to this problem might be to use a weighted connection pool.
I'm experimenting with it in the Elasticsearch client, you can find [here](https://github.com/elastic/elasticsearch-js/blob/weighted-connection-pool/lib/pool/WeightedConnectionPool.js) the implementation, while [here](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.863.3985&rep=rep1&type=pdf) you can find the white paper we took inspiration from.

In this case, we could set the weight of each pipeline based on a heuristic of our choice (latency?), and assign the requests accordingly.

**Comment by mcollina** [comment]
Wow, that is a nice idea.

**Comment by ronag** [comment]
From a quick glance I'm not sure I quite understand how we would apply it here.

**Comment by ronag** [comment]
@delvedor would you be interested in trying to port that into undici.Pool?

**Comment by ronag** [comment]
Another idea:

If any request in the pipeline is serving a response or request with a payload larger than a threshold, label that pipeline as busy and prefer other connections. This is in order to reduce latency where a small request is behind a large on.

@mcollina WDYT?

**Comment by mcollina** [comment]
> Could possible do the same with a timeout threshold, i.e. if the request in the pipeline has run for longer than a period, set it as busy.

This is actually nice. I would do it on-insertion in the pool instead of using terms.

**Comment by ronag** [comment]
Another idea, check if there is a range header and if the range is larger than a treshold stop pipelining further requests.

Pipelining is mainly useful when having lots of small/fast requests. Let's say the TCP + SSL handshake has ~150 ms latency overhead. Any request that takes longer than e.g. 15 seconds (1% overhead) wouldn't gain much (at least in terms of throughput) from pipelining and the Pool would be better of scheduling requests on the next connection.

**Comment by ronag** [comment]
I wonder if we could provide API/events to make this something that can be implemented and experimented with in user land...

**Comment by mcollina** [comment]
+1!
